src/Eloquent/Builder.php

```diff
@@ -183,6 +186,41 @@ public function raw($value = null)
         return $results;
     }
 
+    /**
+     * Attempt to create the record if it does not exist with the matching attributes.
+     * If the record exists, it will be returned.
+     *
+     * @param  array $attributes The attributes to check for duplicate records
+     * @param  array $values     The attributes to insert if no matching record is found
+     */
+    public function createOrFirst(array $attributes = [], array $values = []): Model
+    {
+        // Apply casting and default values to the attributes
+        $instance = $this->newModelInstance($values + $attributes);
+        $values = $instance->getAttributes();
+        $attributes = array_intersect_key($attributes, $values);
+
+        return $this->raw(function (Collection $collection) use ($attributes, $values) {
+            $listener = new FindAndModifyCommandSubscriber();
+            $collection->getManager()->addSubscriber($listener);
+
+            try {
+                $document = $collection->findOneAndUpdate(
+                    $attributes,
+                    ['$setOnInsert' => $values],
```

Comment:
Since `$values` is over-written above, its value here includes the union of fields from the `$attributes` and `$values` parameters. This is a bit redundant, since upserts already apply the criteria to a new document, but I suppose it makes no functional difference.

Just to confirm, what prevents a user from using query syntax in the `$attributes` parameter? It'd logically work for executing `findOneAndUpdate()`, but I imagine it wouldn't play nice with `Builder::newModelInstance()`.

Reply:
Oh, I didn't know the `$filter` argument was used to create the document, good to know.

> Just to confirm, what prevents a user from using query syntax in the `$attributes` parameter? It'd logically work for executing `findOneAndUpdate()`, but I imagine it wouldn't play nice with `Builder::newModelInstance()`.

`$parameters` and `$values` are supposed to be used to create a Model instance. Using a query operator would be incorrect.

Reply:
If `$values` is empty, I get this server error: 
> MongoDB\Driver\Exception\CommandException: Modifiers operate on fields but we found type array instead. For example: {$mod: {<field>: ...}} not {$setOnInsert: []}

If I try to use an empty update, the error is:
> MongoDB\Exception\InvalidArgumentException: Expected update operator(s) or non-empty pipeline for $update

I could add a ternary condition to use `$attribute` as fallback if `$values` is empty, but I do prefer to be explicit and merge all values all the time.

Reply:
> Modifiers operate on fields but we found type array instead.

That could be fixed with an explicit `(object)` cast on the `$setOnInsert` value. Note that server versions prior to 5.0 would still error if that object is empty (this is noted in the [$setOnInsert](https://www.mongodb.com/docs/manual/reference/operator/update/setOnInsert/) docs).

> Expected update operator(s) or non-empty pipeline for $update

Also addressable with an explicit object cast on the value; however, it's secondary to the issue above so not worth looking into.